### PR TITLE
mem: Add faster tiered buffer pool

### PR DESCRIPTION
**Pull request by arjan-bal**
This change adds a new tiered buffer pool that uses power-of-2 tier sizes. It reduces the lookup time for the relevant sizedBufferPool from $O(\log n)$ to $O(1)$, where n is the number of tiers. This creates constant-time lookups independent of the tier count, allowing users to add more tiers without performance overhead.

## Benchmarks

Micro-benchmark that measures only the pool query performance, ignoring the allocation time:
```go
func BenchmarkSearch(b *testing.B) {
	defaultBufferPoolSizes := make([]int, len(defaultBufferPoolSizeExponents))
	for i, exp := range defaultBufferPoolSizeExponents {
		defaultBufferPoolSizes[i] = 1 << exp
	}
	b.Run("pool=Tiered", func(b *testing.B) {
		p := NewTieredBufferPool(defaultBufferPoolSizes...).(*tieredBufferPool)
		for b.Loop() {
			for size := range 1 << 19 {
				// One for get, one for put.
				_ = p.getPool(size)
				_ = p.getPool(size)
			}
		}
	})

	b.Run("pool=BinaryTiered", func(b *testing.B) {
		p := NewBinaryTieredBufferPool(defaultBufferPoolSizeExponents...).(*binaryTieredBufferPool)
		for b.Loop() {
			for size := range 1 << 19 {
				_ = p.poolForGet(size)
				_ = p.poolForPut(size)
			}
		}
	})
}
```
With 5 tiers:
```sh
go test -bench=BenchmarkSearch -count=10 -benchmem | benchstat -col '/pool' -
goos: linux
goarch: amd64
pkg: google.golang.org/grpc/mem
cpu: Intel(R) Xeon(R) CPU @ 2.60GHz
          │   Tiered    │            BinaryTiered             │
          │   sec/op    │   sec/op     vs base                │
Search-48   5.353m ± 2%   2.036m ± 0%  -61.97% (p=0.000 n=10)

          │   Tiered   │          BinaryTiered          │
          │    B/op    │    B/op     vs base            │
Search-48   0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal

          │   Tiered   │          BinaryTiered          │
          │ allocs/op  │ allocs/op   vs base            │
Search-48   0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```
With 9 tiers:
```sh
go test -bench=BenchmarkSearch -count=10 -benchmem | benchstat -col '/pool' -
goos: linux
goarch: amd64
pkg: google.golang.org/grpc/mem
cpu: Intel(R) Xeon(R) CPU @ 2.60GHz
          │   Tiered    │            BinaryTiered             │
          │   sec/op    │   sec/op     vs base                │
Search-48   5.659m ± 0%   2.035m ± 0%  -64.04% (p=0.000 n=10)

          │   Tiered   │          BinaryTiered          │
          │    B/op    │    B/op     vs base            │
Search-48   0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal

          │   Tiered   │          BinaryTiered          │
          │ allocs/op  │ allocs/op   vs base            │
Search-48   0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```

RELEASE NOTES:
* mem: Add faster tiered buffer pool. Use `NewBinaryTieredBufferPool` to create such pools.

### CODE REVIEW

File: mem/buffer_pool.go [hunk 1]
```diff
@@ -109,6 +111,132 @@ func (p *tieredBufferPool) getPool(size int) BufferPool {
...
+// Note that the argument passed to this functions are the powers of 2 of the
+// capacity of the buffers in the pool, not the capacities of the buffers
+// themselves. For example, if you wanted a pool that had buffers with a capacity
+// of 16kb, you would pass 14 as the argument to this function.
+func NewBinaryTieredBufferPool(powerOfTwoExponents ...int) BufferPool {
+	sort.Ints(powerOfTwoExponents)
+
+	// Determine the maximum exponent we need to support.
+	// bits.Len64(math.MaxUint64) is 63.
+	const maxExponent = 63
```
Comment (easwars) [inline mem/buffer_pool.go]: Does this assume a 64-bit machine?

File: mem/buffer_pool.go [hunk 2]
```diff
@@ -109,6 +111,132 @@ func (p *tieredBufferPool) getPool(size int) BufferPool {
...
+	indexOfNextLargestBit := make([]int, maxExponent+1)
+	indexOfPreviousLargestBit := make([]int, maxExponent+1)
+
+	// Initialize with sentinel values
+	for i := range indexOfNextLargestBit {
+		indexOfNextLargestBit[i] = -1
+		indexOfPreviousLargestBit[i] = -1
+	}
```
Comment (easwars) [inline mem/buffer_pool.go]: We could probably use `slices.Repeat([]int{-1}, numElements)` to shorten this.

File: mem/buffer_pool.go [hunk 3]
```diff
@@ -109,6 +111,132 @@ func (p *tieredBufferPool) getPool(size int) BufferPool {
...
+
+	maxCap := 0
+	pools := make([]*sizedBufferPool, 0, len(powerOfTwoExponents))
+
+	for i, exp := range powerOfTwoExponents {
+		if exp > maxExponent || exp < 0 {
+			continue
+		}
```
Comment (easwars) [inline mem/buffer_pool.go]: Are we silently ignoring certain exponents that don't meet our criteria? Should we error out instead?

File: mem/buffer_pool.go [hunk 4]
```diff
@@ -109,6 +111,132 @@ func (p *tieredBufferPool) getPool(size int) BufferPool {
 	return p.sizedPools[poolIdx]
 }
 
+type binaryTieredBufferPool struct {
+	indexOfNextLargestBit     []int
+	indexOfPreviousLargestBit []int
```
Comment (easwars) [inline mem/buffer_pool.go]: It is not very intuitive what these slices represent. Could you please add some comments?

File: mem/buffer_pool.go (hunk 1 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: It works for machines up to 64 bits, i.e. both 32 and 64 bits.

File: mem/buffer_pool.go (hunk 2 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: Nice, wasn't aware of this. Updated.

File: mem/buffer_pool.go (hunk 3 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: Added a comment explaining the reason to ignore the values. 
1. For values greater than 63, it's not possible to allocate such large slices. So ignoring them will not cause observable changes. I changed the param type to `uint8` to help give linter errors if people pass constant values > 256.
2. For values < 1, the buffer sizes would be fractional, e.g.  $`2^{-1}=0.5`$. Since slices are of integer sizes, a sized buffer pool with a fractional size will never get used. I changed the param type to `uint`, making it impossible to pass negative values.

File: mem/buffer_pool.go (hunk 4 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: Added a comment and renamed the slices.

**Comment by arjan-bal** [comment]
> I thought we decided to make it possible for the user to set the default buffer pool for the whole process through an experimental API, and get rid of the existing dial option and the server option. Was your plan to do that in a follow-up?

I was waiting for the author of https://github.com/grpc/grpc-go/pull/8770 to raise a PR for exposing a function to set the default buffer pool. See the second part of https://github.com/grpc/grpc-go/pull/8770#issuecomment-3658968490.

> get rid of the existing dial option and the server option

I'm not sure if we want to do this. Maybe people want to use different buffer pools for each channel.

---
In this PR, I'm just improving the existing buffer pool implementation.

File: mem/buffer_pool.go [hunk 5]
```diff
@@ -109,6 +112,135 @@ func (p *tieredBufferPool) getPool(size int) BufferPool {
...
+	// to return a buffer to the most appropriate pool based on its capacity.
+	exponentToPreviousLargestPoolMap []int
+	sizedPools                       []*sizedBufferPool
+	fallbackPool                     simpleBufferPool
+	maxPoolCap                       int // Optimization: Cache max capacity
+}
+
+// NewBinaryTieredBufferPool returns a BufferPool implementation that uses
+// multiple underlying pools of the given pool sizes. The pool sizes must be
+// powers of 2. This enables O(1) lookup when getting or putting a buffer.
```
Comment (easwars) [inline mem/buffer_pool.go]: This section needs some updating I guess given that we accept exponents as arguments (and the next paragraph also says so).

File: mem/buffer_pool.go (hunk 1 above)
Comment (easwars) [reply mem/buffer_pool.go]: If a value of `63` is passed to this function, `capSize := 1 << exp` will evaluate to a negative number on 64 bit machines, since `capSize` is a signed integer.

Also, how would these bit shift operators (where the exponent is greater than 32 bits) return expected values on 32 bit machines?

Comment (easwars) [reply mem/buffer_pool.go]: I think we should have some tests for the edge cases for both 32 and 64 bit machines.

File: mem/buffer_pool.go (hunk 3 above)
Comment (easwars) [reply mem/buffer_pool.go]: Wouldn't it be better to return an error in such cases instead of ignoring certain exponents?

**Comment by easwars** [comment]
Also, would it make sense to add some of the micro benchmarks that you used as part of this PR? Thanks.

File: mem/buffer_pool.go (hunk 5 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: Updated.

File: mem/buffer_pool.go (hunk 1 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: You're right. I've updated the implementation to use `bits.UintSize` to determine the machine's word size and set that as the maximum tier size. I also added tests that mock `bits.UintSize` to simulate both 32-bit and 64-bit environments.

File: mem/buffer_pool.go (hunk 3 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: Updated to return an error from the constructor.

File: mem/buffer_pool.go [hunk 6]
```diff
@@ -38,22 +41,30 @@ type BufferPool interface {
...
+	14, // 16KB (max HTTP/2 frame size used by gRPC)
+	15, // 32KB (default buffer size for io.Copy)
+	20, // 1MB
 }
 
-var defaultBufferPool BufferPool
+var (
+	defaultBufferPool BufferPool
+	uintSize          = bits.UintSize // use a variable for mocking during tests.
+)
```
Comment (easwars) [inline mem/buffer_pool.go]: Nit: Group all these vars in one block, or don't have a block at all (like you currently do for consts).

File: mem/buffer_pool.go [hunk 7]
```diff
@@ -109,6 +120,131 @@ func (p *tieredBufferPool) getPool(size int) BufferPool {
...
+	indexOfPreviousLargestBit := slices.Repeat([]int{-1}, maxExponent+1)
+
+	maxTier := 0
+	pools := make([]*sizedBufferPool, 0, len(powerOfTwoExponents))
+
+	for i, exp := range powerOfTwoExponents {
+		// Allocating slices of size > 2^maxExponent isn't possible on
+		// maxExponent-bit machines.
+		if int(exp) > maxExponent {
+			return nil, fmt.Errorf("allocating slice of size 2^%d is not possible", exp)
```
Comment (easwars) [inline mem/buffer_pool.go]: Nit: prefix the error string with package name?

File: mem/buffer_pool.go [hunk 8]
```diff
@@ -109,6 +120,131 @@ func (p *tieredBufferPool) getPool(size int) BufferPool {
...
+	// size=16 (0b10000) -> size-1=15 (0b01111) -> bits.Len=4 -> Pool for 2^4
+	// size=17 (0b10001) -> size-1=16 (0b10000) -> bits.Len=5 -> Pool for 2^5
+	querySize := uint(size - 1)
+	poolIdx := b.exponentToNextLargestPoolMap[bits.Len(querySize)]
+
+	return b.sizedPools[poolIdx]
+}
+
+func (b *binaryTieredBufferPool) Put(buf *[]byte) {
+	b.poolForPut(cap(*buf)).Put(buf)
```
Comment (easwars) [inline mem/buffer_pool.go:222]: Can we add a comment here capturing the subtle but important fact that we are passing the capacity of the buffer, and not the size of the buffer to `poolForPut`. If we did the latter, all buffers would eventually move to the smallest pool.

File: mem/buffer_pool_internal_test.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+	defer func() { uintSize = origUintSize }()
+
+	tests := []struct {
+		name      string
+		wordSize  int
+		exponents []uint8
+		wantErr   bool
+	}{
+		{
+			name:      "32-bit valid exponent",
```
Comment (easwars) [inline mem/buffer_pool_internal_test.go]: Nit: Make subtest names more like identifiers. go/go-style/decisions#subtest-names

File: mem/buffer_pool_internal_test.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			uintSize = tt.wordSize
+			pool, err := NewBinaryTieredBufferPool(tt.exponents...)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewBinaryTieredBufferPool() error = %t, wantErr %t", err, tt.wantErr)
+				return
+			}
+			if err == nil {
```
Comment (easwars) [inline mem/buffer_pool_internal_test.go]: Nit: invert the conditional and return early for tests where an error was expected and was seen.

File: mem/buffer_pool_internal_test.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+	b.Run("pool=BinaryTiered", func(b *testing.B) {
+		pool, err := NewBinaryTieredBufferPool(defaultBufferPoolSizeExponents...)
+		if err != nil {
+			b.Fatalf("Failed to create buffer pool: %v", err)
+		}
+		p := pool.(*binaryTieredBufferPool)
+		for b.Loop() {
+			for size := range 1 << 19 {
+				_ = p.poolForGet(size)
+				_ = p.poolForPut(size)
```
Comment (easwars) [inline mem/buffer_pool_internal_test.go:110]: Does it matter that we are not passing capacity here to `poolForPut`?

**Review by easwars — APPROVED** [review]
LGTM modulo minor nits

File: mem/buffer_pool.go (hunk 6 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: Grouped all the vars and consts in a single block.

File: mem/buffer_pool.go (hunk 7 above)
Comment (arjan-bal) [reply mem/buffer_pool.go]: Added the prefix.

File: mem/buffer_pool.go (hunk 8 above)
Comment (arjan-bal) [reply mem/buffer_pool.go:222]: Added a comment.

File: mem/buffer_pool_internal_test.go (hunk 9 above)
Comment (arjan-bal) [reply mem/buffer_pool_internal_test.go]: Done.

File: mem/buffer_pool_internal_test.go (hunk 10 above)
Comment (arjan-bal) [reply mem/buffer_pool_internal_test.go]: Done.

File: mem/buffer_pool_internal_test.go (hunk 11 above)
Comment (arjan-bal) [reply mem/buffer_pool_internal_test.go:110]: It is true that this benchmark doesn't use the buffer's expected capacity, but the results should still be similar. The benchmark intentionally avoids fetching a buffer to ensure we measure only the buffer overhead, excluding allocation time. While we could determine the expected capacity by type-asserting the pool to sizedBufferPool and reading its defaultSize field, this would make the benchmark brittle by relying on a private implementation. Therefore, I am not making the change at this time.

File: mem/buffer_pool.go [hunk 12]
```diff
@@ -109,6 +121,134 @@ func (p *tieredBufferPool) getPool(size int) BufferPool {
...
+}
+
+// NewBinaryTieredBufferPool returns a BufferPool backed by multiple sub-pools.
+// This structure enables O(1) lookup time for Get and Put operations.
+//
+// The arguments provided are the exponents for the buffer capacities (powers
+// of 2), not the raw byte sizes. For example, to create a pool of 16KB buffers
+// (2^14 bytes), pass 14 as the argument.
+func NewBinaryTieredBufferPool(powerOfTwoExponents ...uint8) (BufferPool, error) {
+	slices.Sort(powerOfTwoExponents)
```
[the remark this answers is not in the document]
Comment (arjan-bal) [reply mem/buffer_pool.go]: Added a test for this also.

**Comment by ash2k** [comment]
> I'm not sure if we want to do this. Maybe people want to use different buffer pools for each channel.

Very much agree with this. As a user, I'd prefer to not have any global state at all. I'd prefer all things to be tied to a client or to a server. I can share what I want to share across instances of these by myself. This is a more flexible approach.
